### ISSUE TITLE
Creep is rendered before the glass surface it should be over 

### ISSUE THREAD

**Issue by illwieckz** [DaemonEngine/Daemon#1025]
Creep is rendered before the glass surface it should be over 

![unvanquished_2024-01-13_143531_000](https://github.com/DaemonEngine/Daemon/assets/2311649/98a974f2-e128-4f09-bf77-4c3f50909386)

**Comment by illwieckz** [comment]
This affects any impact:

![unvanquished_2024-01-13_151156_000](https://github.com/DaemonEngine/Daemon/assets/2311649/191ab391-8e3f-440b-b9a8-f499cb2f6578)

**Comment by slipher** [comment]
It's possible that this is the same underlying problem as https://github.com/DaemonEngine/Daemon/issues/906. Either one might be caused by the mark polys being generated insufficiently far in front of the surface.

**Comment by VReaperV** [comment]
This is supposed to be worked around by using polygonOffset, but maybe the factor and offset that we use are wrong.

**Comment by slipher** [comment]
Creep has the `SS_DECAL` sort order so it is drawn before almost all translucent surfaces. And it is treated as a translucent surface, so depth doesn't even matter since the dueling surfaces are both translucent.

So the renderer is working as designed here... but it's an old design based on heuristics that favor batching surfaces with a shared texture to improve speed at the cost of correctness.

With the material system, everything renders as we would like, yay! I think it might ignore fine-grained shader sorting distinctions?

Until [very recently](https://github.com/UnvanquishedAssets/res-buildables_src.dpkdir/commit/ace4237ffb82e7a9db3c316bca28bfca92d5fb24), the creep texture had only 0 or 1 alpha values. So we could have fixed it without changing the sort order by using `depthWrite` in the shader. Now it has alpha that fades out at the edges so we couldn't do that.

**Comment by VReaperV** [comment]
> I think it might ignore fine-grained shader sorting distinctions?

Yeah, it essentially skips this sort when adding surfaces since it just does `R_AddWorldSurfaces()`, and then they either end up in the same material in the order they were received in or implicitly correctly sorted by material.

**Comment by VReaperV** [comment]
This might be caused by the same issue as #1079.

**Comment by slipher** [comment]
> This might be caused by the same issue as #1079.

Sort of. Both are caused by a farther back surface being drawn after one in front. The difference is that in this case the renderer is "required" to draw the closer surface first, according to the `shaderSort_t` ordering. In #1079 the surfaces have the same sort number, so which one is drawn first is decided by some other arbitrary rules.

**Comment by VReaperV** [comment]
> > This might be caused by the same issue as #1079.
> 
> Sort of. Both are caused by a farther back surface being drawn after one in front. The difference is that in this case the renderer is "required" to draw the closer surface first, according to the `shaderSort_t` ordering. In #1079 the surfaces have the same sort number, so which one is drawn first is decided by some other arbitrary rules.

Perhaps this explicit sort should just be ignored? I guess it was used to be able to sort transparents by shader while *generally* avoiding issues like this by explicitly assigning some sort to the shader.

**Comment by slipher** [comment]
I've thought about adding a cvar for sorting the translucent surfaces by distance (assuming we actually have the information to do that?) and ignoring all the other sorting criteria. This would probably be off for lower presets since it makes the renderer slower by preventing merging.

**Comment by VReaperV** [comment]
Heh, that I had the same thought. I'm pretty sure that `R_RecursiveWorldNode()` inherently creates the list of BSP surfaces sorted by distance.
